Site Settings: Add more settings Composing card for Jetpack sites

tyxla opened the pull request:
This PR implements After the Deadline settings for Jetpack sites in the Writing > Composing card. It also moves the current Markdown and Default Post Format settings in separate components in order to keep the main Composing card component clean. It also introduces a `setFieldValue` method to `wrapSettingsForm` that allows manually setting a specific field's value. The PR is part of #9171.

Preview (with collapsed Advanced settings):
![](https://cldup.com/D29VMOgkrO.png)

Preview (with expanded Advanced settings):
![](https://cldup.com/rszM-IjEbJ.png)

To test:

* Checkout this branch
* Go to `/settings/writing/$site` where `$site` is one of your Jetpack sites.
* Verify the default post format field works properly like before.
* Verify the Markdown field is not displayed.
* Play with the new After the Deadline settings (main toggle and sub settings) and verify they are stored and retrieved properly.
* Go to `/settings/writing/$site` where `$site` is one of your WordPress.com sites.
* Verify the default post format field works properly like before.
* Verify the Markdown field works properly like before.
* Verify the After the Deadline settings are not displayed for this site.

Styles here could benefit from #11240 - I can update the PR once it's merged.

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,191 @@
+				<JetpackModuleToggle
+					siteId={ selectedSiteId }
+					moduleSlug="after-the-deadline"
+					label={ translate( 'Check your spelling, style, and grammar.' ) }
```
Hi! I've found a possible matching string that has already been translated **23** times:
`translate( 'Spelling and Grammar' )` *ES Score: 8.14*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,191 @@
+					{ translate( 'English Options' ) }
+				</FormLegend>
+				<FormSettingExplanation>
+					{ translate( 'Enable proofreading for the following grammar and style rules:' ) }
```
Hi! I've found a possible matching string that has already been translated **40** times:
`translate( 'Enable proofreading for the following grammar and style rules when writing posts and pages:' )` *ES Score: 15.58*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,191 @@
+					{ translate( 'Automatic Language Detection' ) }
+				</FormLegend>
+				<FormSettingExplanation>
+					{ translate( 'The proofreader supports English, French, German, Portuguese and Spanish.' ) }
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'The proofreader supports English, French, German, Portuguese, and Spanish. Your <a href="options-general.php">blog language</a> setting is the default proofreading language.' )` *ES Score: 11.38*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,191 @@
+
+				{
+					this.renderToggle( 'onupdate', ! afterTheDeadlineModuleActive, translate(
+						'A post or page is updated'
```
Hi! I've found a possible matching string that has already been translated **47** times:
`translate( 'a post or page is updated' )` *ES Score: 12.39*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,191 @@
+
+				{
+					this.renderToggle( 'onpublish', ! afterTheDeadlineModuleActive, translate(
+						'A post or page is first published'
```
Hi! I've found a possible matching string that has already been translated **45** times:
`translate( 'a post or page is first published' )` *ES Score: 13.52*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

rickybanister commented:
This looks good. I imagine the other PR to standardize the right-margin of toggles will take care of the misalignment here.

The only other thing I noticed was the lack of initial capitals on the first two proofreading settings:

<img width="72" alt="screen shot 2017-02-08 at 10 21 38 am" src="https://cloud.githubusercontent.com/assets/437258/22743571/8f23acdc-ede8-11e6-8b7b-5358c598caa4.png">

All our other labels have an initial capital letter, we should standardize.

rickybanister commented:
To follow up, if we don't like that those two labels start with `a...` we could change them to be plural:

`Posts or pages are first published`
`Posts or pages are updated`

tyxla commented:
> This looks good. I imagine the other PR to standardize the right-margin of toggles will take care of the misalignment here.

Yes - I'll make sure to do any necessary updates/cleanups here once that PR lands.

> To follow up, if we don't like that those two labels start with a... we could change them to be plural:

Even better! I've updated the labels to the one you suggested. 

@rickybanister do you feel there's anything else we should improve, design-wise? Also cc @MichaelArestad for a second opinion.

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,198 @@
+
+				{
+					this.renderToggle( 'onupdate', ! afterTheDeadlineModuleActive, translate(
+						'Posts or pages are updated'
```
Hi! I've found a possible matching string that has already been translated **47** times:
`translate( 'a post or page is updated' )` *ES Score: 7.10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,198 @@
+
+				{
+					this.renderToggle( 'onpublish', ! afterTheDeadlineModuleActive, translate(
+						'Posts or pages are first published'
```
Hi! I've found a possible matching string that has already been translated **45** times:
`translate( 'a post or page is first published' )` *ES Score: 9.02*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

MichaelArestad commented:
@tyxla We should hide the majority of those settings in an "Advanced options" section like we are doing in the `feature/settings-overhaul` branch of Jetpack:

![image](https://cloud.githubusercontent.com/assets/1123119/22783755/794da72e-eecd-11e6-992f-525c88722228.png)

tyxla commented:
@MichaelArestad of course! Thanks for pointing that out. It's now implemented in 202a41a9f61c07a90e527d0bcdd2a5afc36c28f2, I've also updated the previews in the PR description. Feel free to have another 👁 .

MichaelArestad commented on client/my-sites/site-settings/composing/style.scss:
```diff
@@ -0,0 +1,7 @@
+.composing__module-settings.is-indented {
+	margin: 16px 32px 0;
```
Can we drop this in favor of the solution offered here: https://github.com/Automattic/wp-calypso/pull/11240

tyxla replied:
Yes, thanks for pointing that out. I've already added that to the description of the PR, and I'm planning to address it after it's merged.

tyxla replied:
@MichaelArestad this has been implemented in 76aa0071685542ed57d5f37cee65fcde051647f2.

MichaelArestad commented:
Visually, looks :100:

Nice work.

oskosk commented:
Is it possible that some ATD settings still need to be added to the whitelist in `state/jetpack/settings/utils.js`?

tyxla commented:
> Is it possible that some ATD settings still need to be added to the whitelist in state/jetpack/settings/utils.js?

Of course! I totally missed that. Thanks @oskosk, good catch! Added in 65144ec948d730e7b0aa4a283feb3775ac1f792b.

tyxla commented:
This could use a new review @oskosk 😉

oskosk commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,218 @@
+				<div className="composing__info-link-container">
+					<InfoPopover position={ 'left' }>
+						<ExternalLink href={ 'https://jetpack.com/support/spelling-and-grammar/' } target="_blank">
+							{ translate( 'Learn more about After the Deadline' ) }
```
I'm not quite sure if this should read `Learn more about After the Deadline` or `Learn more about Spelling and Grammar`. Although the latter would sound like we're trying to school the user: P. This could be the topic for another PR though.

tyxla replied:
😆 

Well, the module label is "Spelling and Grammer" in Jetpack indeed, so what you say makes sense. In the same time, I kinda agree with you on the teaching the user argument. Perhaps it's best if we get some more opinions on this? @MichaelArestad @rickybanister what do you think about this one?

oskosk commented on client/my-sites/site-settings/composing/index.jsx:
```diff
@@ -90,13 +70,22 @@ Composing.defaultProps = {
+
+		return {
+			jetpackSettingsUISupported: siteSupportsJetpackSettingsUi( state, siteId ),
+			siteIsJetpack: isJetpackSite( state, siteId ),
```
Why backwards? Weren't we using  `props.isJetpackSite` at some other places ?. Well I've just realized out why :P

oskosk commented:
LGTM. Every settings works as expected. The comment I left about the text in the InfoPopover can be handled later. 🚢

rickybanister commented:
Two very minor things we noticed:

It's been decided to remove all the punctuation from labels, so no need for the `.` at the end of the `Check your spelling, style, and grammar`

And last thing—can 'Advanced Options' be aligned with the text label above since those options are children of the parent toggle. Only when the advanced options are related to the section _in general_ would we want it to be fully left-aligned.

tyxla commented:
@rickybanister minor things addressed. A small preview of the updated version: 
![](https://cldup.com/LD9Jqs5u4L.png)

Feel free to have another look.

a8ci18n commented on client/my-sites/site-settings/composing/after-the-deadline.jsx:
```diff
@@ -0,0 +1,220 @@
+				<JetpackModuleToggle
+					siteId={ selectedSiteId }
+					moduleSlug="after-the-deadline"
+					label={ translate( 'Check your spelling, style, and grammar' ) }
```
Hi! I've found a possible matching string that has already been translated **23** times:
`translate( 'Spelling and Grammar' )` *ES Score: 8.15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

rickybanister commented:
@tyxla beautiful, ship it.

folletto commented:
Sorry if I missed this before, but wouldn't be better here to use our foldable card pattern instead of a link to open "Advanced settings"?

Something like this:

![foldable-spelling-card](https://cloud.githubusercontent.com/assets/4389/22925145/8d128e16-f29f-11e6-9a5b-5929d46e77b1.png)

MichaelArestad commented:
@folletto That's been brought up recently by @rickybanister as well. It's something we're definitely exploring, but this works and is good to merge for now.
